internal/querycoordv2/task/scheduler.go

```diff
@@ -0,0 +1,683 @@
+}
+
+func (queue *taskQueue) Cap() int {
+	return queue.cap
+}
+
+func (queue *taskQueue) Add(task Task) bool {
+	if queue.Len() >= queue.Cap() {
+		return false
+	}
+
+	queue.buckets[task.Priority()] = append(queue.buckets[task.Priority()], task)
+	return true
+}
+
+func (queue *taskQueue) Remove(task Task) {
+	bucket := &queue.buckets[task.Priority()]
+
+	for i := range *bucket {
+		if (*bucket)[i].ID() == task.ID() {
+			*bucket = append((*bucket)[:i], (*bucket)[i+1:]...)
+			break
+		}
+	}
+}
+
+// Range iterates all tasks in the queue ordered by priority from high to low
+func (queue *taskQueue) Range(fn func(task Task) bool) {
+	for priority := len(queue.buckets) - 1; priority >= 0; priority-- {
+		for i := range queue.buckets[priority] {
+			if !fn(queue.buckets[priority][i]) {
+				return
+			}
+		}
+	}
+}
+
+type Scheduler interface {
+	Add(task Task) error
+	Dispatch(node int64)
+	RemoveByNode(node int64)
+	GetNodeSegmentDelta(nodeID int64) int
+	GetNodeChannelDelta(nodeID int64) int
+}
+
+type taskScheduler struct {
+	rwmutex     sync.RWMutex
+	ctx         context.Context
+	executor    *Executor
+	idAllocator func() UniqueID
+
+	distMgr   *meta.DistributionManager
+	meta      *meta.Meta
+	targetMgr *meta.TargetManager
+	broker    meta.Broker
+	nodeMgr   *session.NodeManager
+
+	tasks        UniqueSet
+	segmentTasks map[replicaSegmentIndex]Task
+	channelTasks map[replicaChannelIndex]Task
+	processQueue *taskQueue
+	waitQueue    *taskQueue
+}
+
+func NewScheduler(ctx context.Context,
+	meta *meta.Meta,
+	distMgr *meta.DistributionManager,
+	targetMgr *meta.TargetManager,
+	broker meta.Broker,
+	cluster session.Cluster,
+	nodeMgr *session.NodeManager) *taskScheduler {
+	id := int64(0)
+	return &taskScheduler{
+		ctx:      ctx,
+		executor: NewExecutor(meta, distMgr, broker, targetMgr, cluster, nodeMgr),
+		idAllocator: func() UniqueID {
+			id++
+			return id
+		},
+
+		distMgr:   distMgr,
+		meta:      meta,
+		targetMgr: targetMgr,
+		broker:    broker,
+		nodeMgr:   nodeMgr,
+
+		tasks:        make(UniqueSet),
+		segmentTasks: make(map[replicaSegmentIndex]Task),
+		channelTasks: make(map[replicaChannelIndex]Task),
+		processQueue: newTaskQueue(taskPoolSize),
+		waitQueue:    newTaskQueue(taskPoolSize * 10),
+	}
+}
+
+func (scheduler *taskScheduler) Add(task Task) error {
+	scheduler.rwmutex.Lock()
+	defer scheduler.rwmutex.Unlock()
+
+	err := scheduler.preAdd(task)
+	if err != nil {
+		return err
+	}
+
+	task.SetID(scheduler.idAllocator())
+	scheduler.tasks.Insert(task.ID())
+	log := log.With(
+		zap.Int64("task", task.ID()),
+		zap.Int64("collection", task.CollectionID()),
+		zap.Int32("type", int32(GetTaskType(task))),
+		zap.Int32("priority", task.Priority()),
+		zap.Int64("source", task.SourceID()),
+	)
+	switch task := task.(type) {
+	case *SegmentTask:
+		index := replicaSegmentIndex{task.ReplicaID(), task.segmentID}
+		scheduler.segmentTasks[index] = task
+		log.Info("task added", zap.Int64("segment", task.SegmentID()))
```

Comment:
need to log detailed info about the task.
1) which node
2) growing or shrink
3) probably need to log at checker about why this task is generated

Reply:
> 

I add a function to serialize task to string. No.3 will print too many logs and we disable it.